src/sink/nats.rs

```diff
@@ -0,0 +1,324 @@
+// Copyright 2020-2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(not(tarpaulin_include))]
+
+use std::iter::FromIterator;
+use std::time::Instant;
+
+use crate::sink::prelude::*;
+use async_channel::{bounded, Receiver};
+use async_nats::Connection as NatsConnection;
+use async_nats::Headers;
+use async_nats::Options as NatsOptions;
+use halfbrown::HashMap;
+use tremor_pipeline::OpMeta;
+
+// struct containing connection options
+#[derive(Clone, Debug, Deserialize, Default, PartialEq)]
+pub struct ConnectOptions {
+    token: Option<String>,
+    username: Option<String>,
+    password: Option<String>,
+    credentials_path: Option<String>,
+    cert_path: Option<String>,
+    key_path: Option<String>,
+    name: Option<String>,
+    #[serde(default = "bool::default")]
+    echo: bool,
+    max_reconnects: Option<usize>,
+    reconnect_buffer_size: Option<usize>,
+    #[serde(default = "bool::default")]
+    tls: bool,
+    root_cert: Option<String>,
+}
+
+impl ConnectOptions {
+    pub fn generate(&self) -> Result<NatsOptions> {
+        let mut options = None;
+        if let Some(token) = &self.token {
+            options = Some(NatsOptions::with_token(token.as_str()));
+        }
+        if let Some(username) = &self.username {
+            if let Some(password) = &self.password {
+                options = Some(NatsOptions::with_user_pass(
+                    username.as_str(),
+                    password.as_str(),
+                ));
+            }
+        }
```

Comment:
This can likely be de-nested using Option::zip (I learned that today :D)
```suggestion
        if let Some((username, password)) = &self.username.zip(&self.password) {
            options = Some(NatsOptions::with_user_pass(
                &username,
                &password,
            ));
        }
```
(some for the `cert_client` version)